extjson.go

```diff
@@ -0,0 +1,1694 @@
+		strLength := len(out) - lengthPos - 4
+		overwriteLength(out, lengthPos, strLength)
+	case ',':
+		// Maybe followed by $scope
+		err = d.readQuoteStart()
+		if err != nil {
+			return nil, err
+		}
+		err = d.readSpecificKey(jsonScope)
+		if err != nil {
+			return nil, err
+		}
+
+		// We know it's code w/ scope: add additional length bytes for the
+		// Cstring, then convert the scope object.
+		overwriteTypeByte(out, typeBytePos, bsonCodeWithScope)
+		strLengthPos := len(out)
+		out = append(out, emptyLength...)
+		out = append(out, codeCString...)
+		strLength := len(out) - strLengthPos - 4
+		overwriteLength(out, strLengthPos, strLength)
+
+		err = d.readCharAfterWS('{')
+		if err != nil {
+			return nil, err
+		}
+		out, err = d.convertObject(out, topContainer)
+		if err != nil {
+			return nil, err
+		}
+
+		// BSON code w/scope length is total length including length bytes
+		cwsLength := len(out) - lengthPos
+		overwriteLength(out, lengthPos, cwsLength)
+
+		// Must end with document terminator
+		err = d.readObjectTerminator()
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, d.parseError(ch, "expected value separator or end of object")
+	}
+
+	return out, nil
+}
+
+// convertDate starts after the `"$date"` key.  The value might be
+// an ISO-8601 string or might be a $numberLong object.
+func (d *Decoder) convertDate(out []byte) ([]byte, error) {
+	// Consume ':'
+	err := d.readNameSeparator()
+	if err != nil {
+		return nil, err
+	}
+
+	ch, err := d.readAfterWS()
+	if err != nil {
+		return nil, err
+	}
+	switch ch {
+	case '"':
+		// Shortest ISO-8601 is `YYYY-MM-DDTHH:MM:SSZ` (20 chars); longest is
+		// `YYYY-MM-DDTHH:MM:SS.sss+HH:MM` (29 chars).  Plus we need the closing
+		// quote.  Peek a little further in case extra precision is given
+		// (counter to the spec).
+		buf, err := d.peekBoundedQuote(21, 48)
+		if err != nil {
+			return nil, err
+		}
+		epochMillis, err := parseISO8601toEpochMillis(buf)
+		if err != nil {
+			return nil, fmt.Errorf("parse error: %v", err)
+		}
+		_, _ = d.json.Discard(len(buf) + 1)
+		var x [8]byte
+		xs := x[0:8]
+		binary.LittleEndian.PutUint64(xs, uint64(epochMillis))
+		out = append(out, xs...)
+	case '{':
+		err = d.readQuoteStart()
+		if err != nil {
+			return nil, err
+		}
+		err = d.readSpecificKey(jsonNumberLong)
+		if err != nil {
+			return nil, err
+		}
+		// readSpecificKey eats ':' but convertNumberLong wants it so unread it
+		_ = d.json.UnreadByte()
+		out, err = d.convertNumberLong(out)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, d.parseError(ch, "invalid value for $date")
+	}
+
+	// Must end with document terminator
+	err = d.readObjectTerminator()
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// convertType starts after the opening quote of the `"$type"` key.  We need to
+// distinguish between Extended JSON $type or something else.  If
+// we can peek far enough, we can check with regular expresssions.
```

Comment:
[nit] move this comment to the line before the element it describes (`convertType`)